seqr/utils/middleware.py

```diff
@@ -56,6 +61,70 @@ def __init__(self, errors, warnings=None):
 
 ERROR_LOG_EXCEPTIONS = {InvalidIndexException}
 
+
+class CheckProgrammaticAccessMiddleware(MiddlewareMixin):
+
+    def process_request(self, request):
+        func, _, _ = resolve(request.path)
+        request.programmatic_access = isinstance(func, ProgrammaticAccess)
+
+
+class DisableCSRFProgrammaticAccessMiddleware(MiddlewareMixin):
+
+    def process_view(self, request, callback, callback_args, callback_kwargs):
+        assert hasattr(request, 'programmatic_access'), (
+            'The seqr DisableCSRFProgrammaticAccess middleware requires '
+            'CheckProgrammaticAccessMiddleware middleware to be installed. '
+            'Edit your MIDDLEWARE setting to insert '
+            '"seqr.utils.middleware.CheckProgrammaticAccessMiddleware" before '
+            '"seqr.utils.middleware.DisableCSRFProgrammaticAccessMiddleware".'
+        )
+
+        if request.programmatic_access:
+            # only exempt CSRF if it's a programmatic access route
+            callback.csrf_exempt = True
+            # alternative
+            # request._dont_enforce_csrf_checks = True
+
+
+class BearerAuth(MiddlewareMixin):
+
+    def process_request(self, request):
+
+        assert hasattr(request, 'programmatic_access'), (
+            'The seqr BearerAuth middleware requires '
+            'CheckProgrammaticAccessMiddleware middleware to be installed. '
+            'Edit your MIDDLEWARE setting to insert '
+            '"seqr.utils.middleware.CheckProgrammaticAccessMiddleware" before '
+            '"seqr.utils.middleware.BearerAuth".'
+        )
+
+        if request.programmatic_access:
+
+            assert SOCIAL_AUTH_GOOGLE_OAUTH2_KEY
+
+            authorization_value = request.META.get('HTTP_AUTHORIZATION', '')
+            if not authorization_value.startswith('Bearer'):
+                raise PermissionDenied('Expected Bearer token authorization for programmatic route')
+
+            token = authorization_value.split(' ', maxsplit=1)[-1]
+
+            # Specify the CLIENT_ID of the app that accesses the backend:
+            try:
+                idinfo = id_token.verify_oauth2_token(token, requests.Request(), SOCIAL_AUTH_GOOGLE_OAUTH2_KEY)
```

Comment:
I understand the current OAuth setup is already tied to Google.  This makes it even more so and probably should be renamed to GoogleBearerAuth or something.

Reply:
Great suggestion! I've refactored this to a `GoogleBearerAuth` middleware - with a more generic `BearerAuth` base to allow for easier extensibility.